openedx/features/content_tagging/models.py

```diff
@@ -0,0 +1,163 @@
+        cls,
+        taxonomy: Taxonomy,
+    ) -> models.QuerySet:
+        """
+        Returns the org "owners" on the given taxonomy.
+
+        If the returned list is empty, then the returned taxonomy is enabled for all organizations.
+        """
+        return cls.objects.filter(
+            taxonomy=taxonomy,
+            rel_type=cls.RelType.OWNER,
+        )
+
+    @classmethod
+    def is_owner(cls, taxonomy: Taxonomy, org_short_name: str) -> bool:
+        """
+        Returns True if there's an "owner" relationship between the given taxonomy and an org with the given short name.
+
+        Or, if no "owner" relationships exist for the given taxonomy, then all orgs are enabled for that taxonomy.
+        """
+        org_owners = cls.get_owners(taxonomy)
+        if not org_owners.exists():
+            return True
+        org_owners = org_owners.filter(org__short_name=org_short_name)
+        return org_owners.exists()
+
+
+class CourseObjectTagMixin:
+    """
+    Mixin for ObjectTag that accepts course keys as object IDs.
+    """
+
+    @classmethod
+    def valid_for(cls, object_id: str = None, **kwargs) -> bool:
+        """
+        Returns True if the given object_id is a valid CourseKey.
+        """
+        if not object_id:
+            return False
+
+        try:
+            CourseKey.from_string(object_id)
+            return True
+        except InvalidKeyError:
+            return False
+
+    def _check_object(self):
+        """
+        Returns True if this CourseObjectTag has a valid course key.
+        """
+        try:
+            course_key = CourseKey.from_string(self.object_id)
+        except InvalidKeyError:
+            return False
+
+        # ...and the course must use an org that's enabled for this taxonomy.
+        if not TaxonomyOrg.is_owner(self.taxonomy, course_key.org):
+            return False
+
+        return True
+
+
+class BlockObjectTagMixin:
+    """
+    Mixin for ObjectTag that accepts usage keys as object IDs.
+    """
+
+    @classmethod
+    def valid_for(cls, object_id: str = None, **kwargs) -> bool:
+        """
+        Returns True if the given object_id is a valid UsageKey.
+        """
+        if not object_id:
+            return False
+
+        try:
+            UsageKey.from_string(object_id)
+            return True
+        except InvalidKeyError:
+            return False
+
+    def _check_object(self):
+        """
+        Returns True if this BlockObjectTag has a valid usage key,
+        and an org that's enabled for its taxonomy.
+        """
+        try:
+            usage_key = UsageKey.from_string(self.object_id)
+        except InvalidKeyError:
+            return False
+
+        # ...and the course must use an org that's enabled for this taxonomy.
+        if not TaxonomyOrg.is_owner(self.taxonomy, usage_key.org):
+            return False
+
+        return True
+
+
+class OpenCourseObjectTag(CourseObjectTagMixin, OpenObjectTag):
```

Comment:
@pomegranited Correct me if I'm wrong, but in this case would only the `valid_for` of `CourseObjectTagMixin` be called, right? The `valid_for` of OpenObjectTag would not be called

Reply:
Yes.

Reply:
In that case, wouldn't it be nice to have both validations called? Since the validations are not being joined, but are being crushed with each other

Reply:
@ChrisChV You're correct -- it should be calling the subclasses. But python inheritance doesn't work well with class methods, so your comment made me re-examine the `valid_for` and `is_valid` logic.. there was so much duplication there! So I [refactored oel_tagging](https://github.com/open-craft/edx-platform/commit/74a51e203e3a65e191fc91deb5e97616e4d4e7dd) and [the ObjectTag subclasses here](https://github.com/openedx/edx-platform/pull/32661/commits/74a51e203e3a65e191fc91deb5e97616e4d4e7dd) to remove the class method `valid_for`, so we can use normal subclass overrides and super class calls where needed.

Does this work better for you?

Reply:
@pomegranited Yes! Looks really good! :+1: